docs: document terraform support (#1037)
edgelesssys/constellation: docs/docs/workflows/create.md

*create* stores your cluster's state into a [`terraform.tfstate`](../architecture/orchestration.md#cluster-creation-process) file in your workspace.

</tabItem>
<tabItem value="terraform" label="Terraform">

:::caution

> There is no extensive input validation with Terraform. If unsure, please use the Constellation CLI instead.

Reviewer: how much effort would it be to add input validation?

Author: This mostly depends on how lax we want to be in regards to the instance type aka if we were to allow only a small subset of (C)VM types, then we probably can add validation for this. 
Apart from that, I think the validation is actually quite similar. I also just added some more defaults for AWS.

Reviewer: So I guess with the additions you've made, you still feel this warning is needed and helpful for the reader, right?

Author: I'm not 100% sure, but since Terraform is the "expert" way, I removed the warning for now. If we see common errors we can add it again.

:::

> Constellation also supports managing the infrastructure via Terraform. This allows for an easier GitOps integration as well as meeting regulatory requirements.
> Since the Constellation CLI also uses Terraform under the hood, the same Terraform files can be reused.

Reviewer: ```suggestion
Constellation supports managing the infrastructure via Terraform. This allows for an easier GitOps integration as well as meeting regulatory requirements.
Since the Constellation CLI also uses Terraform under the hood, you can reuse the same Terraform files.
```

> For now, please try to refrain from changing the Terraform resource definitions, as Constellation depends on certain components.

Reviewer: ```suggestion
For now, please refrain from changing the Terraform resource definitions, as Constellation is tightly coupled to them.
```
maybe this is more clear?

> Download the Terraform files for the selected CSP from our [GitHub repository](https://github.com/edgelesssys/constellation/tree/main/cli/internal/terraform/terraform).

Reviewer: ```suggestion
Download the Terraform files for the selected CSP from the [GitHub repository](https://github.com/edgelesssys/constellation/tree/main/cli/internal/terraform/terraform).
```

> Download the Terraform files for the selected CSP from our [GitHub repository](https://github.com/edgelesssys/constellation/tree/main/cli/internal/terraform/terraform).

Reviewer: Do I have to download all files including subdirectories? This is tedious on the GitHub UI. Any idea for a better UX here? (I once used [curl for this](https://github.com/edgelesssys/constellation/tree/main/cli/internal/terraform/terraform/iam/aws), but with subdirectories it's more complicated.)

Author: Yes, basically every time I had the problem I resorted to cloning the entire repository instead. 

We could show those steps. From experience: In the one case we had, the user was able to do this step on their own.

Reviewer: Ok, probably not worth to extend this then.

Create a `terraform.tfvars` file.
There, define all needed variables found in `variables.tf` using the values from the `constellation-config.yaml`.

To find the image reference for your CSP and region, execute:

```bash
CONSTELL_VER=vX.Y.Z
curl -s https://cdn.confidential.cloud/constellation/v1/ref/-/stream/stable/$CONSTELL_VER/image/info.json | jq
```

> Now, initialize and apply Terraform to create the configured infrastructure:

Reviewer: ```suggestion
Initialize and apply Terraform to create the configured infrastructure:
```

```bash
terraform init
terraform apply
```

> The initialization depends on the already created `constellation-config.yaml` and the `constellation-id.json`.
> Therefore, create the `constellation-id.json` using the output from the Terraform state and the `constellation-conf.yaml`:

Reviewer: ```suggestion
The Constellation [init step](#the-init-step) requires the already created `constellation-config.yaml` and the `constellation-id.json`.
Create the latter using the output from the Terraform state and the `constellation-conf.yaml`:
```
I was confused because I first related "the initialization" to `terraform init`, so I suggest to clarify this a bit

Author: Changed `latter` to `constellation-id.json` so one does not need to backtrack to the sentence before.

```bash
CONSTELL_IP=$(terraform output ip)
CONSTELL_INIT_SECRET=$(terraform output initSecret | jq -r | tr -d '\n' | base64)
CONSTELL_CSP=$(cat constellation-conf.yaml | yq ".provider | keys | .[0]")
jq --null-input --arg cloudprovider "$CONSTELL_CSP" --arg ip "$CONSTELL_IP" --arg initsecret "$CONSTELL_INIT_SECRET" '{"cloudprovider":$cloudprovider,"ip":$ip,"initsecret":$initsecret}' > constellation-id.json
```

</tabItem>

> </tabs>

Reviewer: If we create a cluster with these steps, do we have the same files in the workspace as if we used the create command? In particular, will the terminate command work?

Reviewer: Then we need to refer to the Terminate workflow from the "Terminate your cluster" section on the First steps page.

> </tabs>

Reviewer: And regarding terminate, what's the plan for this? Do this later?

Author: No, I just forgot, that we have a separate `terminate.md`. Will add this now.
